### PR TITLE
Adds Support to configure hyper-v disk block size

### LINKED ISSUE #4549: Hyper-V Builder with Linux does not use recommended VHDX settings
According to Microsoft's [best practices for Hyper-V + Linux](https://technet.microsoft.com/en-us/windows-server-docs/compute/hyper-v/best-practices-for-running-linux-on-hyper-v), the VHDX used for a Linux VM in Hyper-V should have a BlockSize of 1 MiB. However, packer is creating VHDX's with the default BlockSize of 32 MiB. From my own tests, I have confirmed that using the default 32 MiB setting significantly increases the real disk space usage of the dynamic VHDX. (I'm working on a test case to show the difference.)

Packer version: Packer v0.12.2
Host platform: Windows 10 Pro Insider Preview (build 14986.1001), 64 bit
To reproduce:
Create any Linux+Hyper-V base image. (Examples [here](https://github.com/taliesins/packer-baseboxes
) if you need a template)

Examine the output VHDX in PowerShell:
```
PS C:\Users\Evan\AppData\Local\Temp\packerhv016796879> Get-VHD .\ubuntu-xenial.vhdx


ComputerName            : XYZ
Path                    : c:\users\evan\appdata\local\temp\packerhv016796879\ubuntu-xenial.vhdx
VhdFormat               : VHDX
VhdType                 : Dynamic
FileSize                : 2654994432
Size                    : 22481469440
MinimumSize             : 22480437760
LogicalSectorSize       : 512
PhysicalSectorSize      : 4096
BlockSize               : 33554432
ParentPath              :
DiskIdentifier          : 78A0D4B6-155C-4621-B947-C6749807581B
FragmentationPercentage : 7
Alignment               : 1
Attached                : True
DiskNumber              :
Number                  :
```

Notice that the BlockSize attribute is 33554432 bytes, or 32 MiB.

Compare to a VHDX created with the recommended BlockSize of 1 MiB:

```
PS C:\Users\Evan\AppData\Local\Temp\packerhv016796879> New-VHD -path ./test.vhdx -SizeBytes 10GB -dynamic -BlockSizeBytes 1MB

ComputerName            : XYZ
Path                    : C:\Users\Evan\AppData\Local\Temp\packerhv016796879\test.vhdx
VhdFormat               : VHDX
VhdType                 : Dynamic
FileSize                : 4194304
Size                    : 10737418240
MinimumSize             :
LogicalSectorSize       : 512
PhysicalSectorSize      : 4096
BlockSize               : 1048576
ParentPath              :
DiskIdentifier          : 598D6DAC-8385-4FCE-84CE-D4D8CDDC25A0
FragmentationPercentage : 0
Alignment               : 1
Attached                : False
DiskNumber              :
Number                  :
```

Debug log (only has up to the point where the VM has been booted, but that should be enough)
https://gist.github.com/evancox10/b0982f0c6f9c1b774d5d1bb19de3f2fd

### PR DESCRIPTION
**Pull request by synax**
_Note: I've merged and tested the change implemented by @evancox10 myself, since I need this feature and there was no progress after September 2017. The credit for implementing this goes to @evancox10._

This adds the option to specify the VHD block size for hyper-v builders. This is particularly required for Linux VM disk size optimization as described here: [tuning-linux-file-systems-on-dynamic-vhdx-files](https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/best-practices-for-running-linux-on-hyper-v#tuning-linux-file-systems-on-dynamic-vhdx-files). The default value is 32 MiB, Microsofts recommendation is 1 MiB.

example config using the new property:

```json
{
"variables": {
"vm_name": "ubuntu-xenial",
"cpu": "2",
"ram_size": "1024",
"disk_size": "21440",
"iso_url": "http://releases.ubuntu.com/16.04/ubuntu-16.04.1-server-amd64.iso",
"iso_checksum_type": "sha1",
"iso_checksum": "DE5EE8665048F009577763EFBF4A6F0558833E59"
},
"builders": [
{
"vm_name":"{{user vm_name}}",
"type": "hyperv-iso",
"disk_size": "{{user disk_size}}",
**"disk_block_size": "1",**
"guest_additions_mode": "disable",
"iso_url": "{{user iso_url}}",
"iso_checksum_type": "{{user iso_checksum_type}}",
"iso_checksum": "{{user iso_checksum}}",
"communicator":"ssh",
"ssh_username": "packer",
"ssh_password": "packer",
"ssh_timeout" : "4h",
"http_directory": "./",
"boot_wait": "5s",
"boot_command": [
"",
"set gfxpayload=1024x768",
"linux /install/vmlinuz ",
"preseed/url=http://{{.HTTPIP}}:{{.HTTPPort}}/hyperv-taliesins.cfg ",
"debian-installer=en_US auto locale=en_US kbd-chooser/method=us ",
"hostname={{.Name}} ",
"fb=false debconf/frontend=noninteractive ",
"keyboard-configuration/modelcode=SKIP keyboard-configuration/layout=USA ",
"keyboard-configuration/variant=USA console-setup/ask_detect=false ",
"initrd /install/initrd.gz",
"boot"
],
"shutdown_command": "echo 'packer' | sudo -S -E shutdown -P now",
"ram_size": "{{user ram_size}}",
"cpu": "{{user cpu}}",
"generation": 2,
"enable_secure_boot": true
}
]
}
```

This could be documented as follows on the hyperv-vmcx and hyperv-iso pages:

disk_block_size (string) - The block size of the VHD to be created. Recommended disk block size for Linux hyper-v guests is 1 MiB. This defaults to "32 MiB".

Closes #4549

### CODE REVIEW

**Comment by mwhooker** [comment]
Thanks for this! also needs a `make fmt`

**Comment by synax** [comment]
Did some more testing. It really looks like using 1 MiB block size causes the linux box to be about 50% less in size. 👍

**Comment by SwampDragons** [comment]
Sorry to do this, but I haven't had a chance to test and review this yet :( Going to push it off until the next release.  Meanwhile, can you write some documentation https://github.com/hashicorp/packer/blob/master/website/source/docs/builders/hyperv-vmcx.html.md and https://github.com/hashicorp/packer/blob/master/website/source/docs/builders/hyperv-iso.html.md for this new feature?

**Comment by synax** [comment]
No worries, I will add the documentation asap.

**Comment by synax** [comment]
I've added the documentation to the hyper-v ISO builder page.

File: common/powershell/hyperv/hyperv.go [hunk 1]
```diff
@@ -53,6 +53,22 @@ $ip
 	return cmdOut, err
 }
 
+//func CreateVirtualHardDiskDrive(vmName string, diskPath string, diskSize int64, diskBlockSize int64, generation uint) (uint, uint, error) {
```
Comment (evancox10) [inline common/powershell/hyperv/hyperv.go]: You probably want to delete this part. This is some scratch code from when I thought about creating VMs with a separate function to make adding additional VHDs easier. Looks like this feature was implemented in https://github.com/hashicorp/packer/pull/5941

File: common/powershell/hyperv/hyperv.go [hunk 2]
```diff
@@ -872,16 +890,16 @@ Hyper-V\Get-VMNetworkAdapter -VMName $vmName | Hyper-V\Connect-VMNetworkAdapter
 	return err
 }
 
-func AddVirtualMachineHardDiskDrive(vmName string, vhdRoot string, vhdName string, vhdSizeBytes int64, controllerType string) error {
+func AddVirtualMachineHardDiskDrive(vmName string, vhdRoot string, vhdName string, vhdSizeBytes int64, vhdBlockSize int64, controllerType string) error {
 
 	var script = `
-param([string]$vmName,[string]$vhdRoot, [string]$vhdName, [string]$vhdSizeInBytes, [string]$controllerType)
+param([string]$vmName,[string]$vhdRoot, [string]$vhdName, [string]$vhdSizeInBytes,[string]$vhdBlockSizeInBize [string]$controllerType)
```
Comment (evancox10) [inline common/powershell/hyperv/hyperv.go]: Minor typo here and on line 898: `$vhdBlockSizeInBize` should be `$vhdBlockSizeInBytes`

**Comment by synax** [comment]
I think I found the issue. I've retested a Gen 1 VM and it worked for me. Can you please retest?

**Pull request merged**